silkworm/execution/domain_state.cpp

```diff
@@ -0,0 +1,167 @@
+/*
+   Copyright 2024 The Silkworm Authors
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+#include "domain_state.hpp"
+
+#include <silkworm/core/common/util.hpp>
+#include <silkworm/db/state/accounts_domain.hpp>
+#include <silkworm/db/state/code_domain.hpp>
+#include <silkworm/db/state/schema_config.hpp>
+#include <silkworm/db/state/storage_domain.hpp>
+
+namespace silkworm::execution {
+
+using namespace db::state;
+using namespace datastore;
+
+// TODO:
+//  - implement: state_root_hash, canonize_block, decanonize_block (optional)
+//  - implement insert_call_traces (mandatory)
+//  - add begin_txn method (replacing begin_block?)
+//  - insert_receipts should write to domain receipts table db::state::kDomainNameReceipts
+//  - add buffer saving previous steps for values in accounts, code and storage domains - for updates
+//  - extend transaction to include txn_id, or
+//  - add base_txn_id to block and get_txn_by_id method
+
+std::optional<Account> DomainState::read_account(const evmc::address& address) const noexcept {
+    AccountsDomainGetLatestQuery query{database_, tx_, state_repository_};
+    auto result = query.exec(address);
+    if (result) {
+        return std::move(result->value);
+    }
+    return std::nullopt;
+}
+
+ByteView DomainState::read_code(const evmc::address& address, const evmc::bytes32& /*code_hash*/) const noexcept {
+    CodeDomainGetLatestQuery query{database_, tx_, state_repository_};
+    auto result = query.exec(address);
+    if (result) {
+        return std::move(result->value);
```

Comment:
FYI there are some upcoming pending changes regarding this method, see: local_state.cpp here - https://github.com/erigontech/silkworm/pull/2680/files#diff-479b908db1d5e43fd1f0249a9c003f3b9a60d3762dd8c79ab70114c9fa76bedbR47